### PR TITLE
Implement escaped-character handling in Utf8JsonReader and JsonElement

### LINKED ISSUE #34690: Add (Try)GetDateTime(Offset) to Utf8JsonReader and JsonDocument
We should be consistent regarding what .NET types the `Utf8JsonReader`, `Utf8JsonWriter`, and `JsonDocument` support.

Since there is strong need for adding DateTime support (and the writer allows users to write one), we should add the following APIs to the reader and document.

**Add:**
```C#
public ref partial struct Utf8JsonReader
{
   // These throw InvalidOperationException if there is a token type mismatch (not a string)
   // Throws format exception if the string is not a DateTime (or DateTimeOffset)
   public DateTime GetDateTime();
   public DateTimeOffset GetDateTimeOffset();

   // These throw InvalidOperationException if there is a token type mismatch (not a string)
   public bool TryGetDateTime(out DateTime value);
   public bool TryGetDateTimeOffset(out DateTimeOffset value);
}
public readonly partial struct JsonElement
{
   public DateTime GetDateTime();
   public DateTimeOffset GetDateTimeOffset();

   public bool TryGetDateTime(out DateTime value);
   public bool TryGetDateTimeOffset(out DateTimeOffset value);
}
```

For consistency, we should remove Write(Guid) from the writer unless a customer requests it. If we want to keep Guid support, add it to the reader/document as well.

**Remove:**
```C#
public ref partial struct Utf8JsonWriter
{
        public void WriteString(System.ReadOnlySpan<byte> utf8PropertyName, System.Guid value, bool escape = true) { }
        public void WriteString(System.ReadOnlySpan<char> propertyName, System.Guid value, bool escape = true) { }
        public void WriteString(string propertyName, System.Guid value, bool escape = true) { }
        public void WriteStringValue(System.Guid value) { }
}
```

cc @bartonjs, @steveharter, @JamesNK

### LINKED ISSUE #36202: Implement escaped-character handling in (Try)GetDateTime(Offset) methods of the Utf8JsonReader
Similar to the implementation for `(Try)GetGuid` (https://github.com/dotnet/corefx/pull/36122), we want these methods to handle escaped characters.

### PR DESCRIPTION
**Pull request by layomia**
Implement escaped-character handling in (Try)GetDateTime(Offset) & (Try)GetGuid in Utf8JsonReader and JsonElement

Fixes #36202
Fixes #34690

Also, cap the number of fraction digits the Utf8JsonReader can parse to 9.

Addition: This PR also adds (Try)GetGuid to JsonElement

### CODE REVIEW

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs [hunk 1]
```diff
@@ -465,8 +465,33 @@ internal bool TryGetValue(int index, out DateTime value)
...
-                segment.Length == bytesConsumed)
+            DateTime tmp;
+            int bytesConsumed;
+
+            if (row.HasComplexChildren)
+            {
+                int backslash = segment.IndexOf(JsonConstants.BackSlash);
+                Debug.Assert(backslash != -1);
+
+                Span<byte> segmentUnescaped = stackalloc byte[segment.Length];
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs]: The length needs to be validated as in-range before a stackalloc.

Comment (GrabYourPitchforks) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs]: For reference, the correct pattern would look something like this:

```cs
Span<byte> theSpan = ((uint)buffer.Length <= 128 /* or whatever const */)
    ? stackalloc byte[buffer.Length]
    : new byte[buffer.Length];
```

File: src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs [hunk 2]
```diff
@@ -248,27 +249,20 @@ private static int LocateFirstFoundByte(ulong match)
                                                0x02ul << 40 |
                                                0x01ul << 48) + 1;
 
-        public static bool TryGetEscapedGuid(ReadOnlySpan<byte> span, out Guid value)
+        public static void GetValueSequenceAsSpan(ReadOnlySequence<byte> valueSequence, out ReadOnlySpan<byte> span)
```
Comment (GrabYourPitchforks) [inline src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs]: What is the intended behavior of this method? The current behavior is that it copies some data within the local frame's buffer, outs a pointer to the local frame, then allows the local frame to unwind. Once the method returns `span` will always point to uninitialized memory.

File: src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs [hunk 3]
```diff
@@ -248,27 +249,20 @@ private static int LocateFirstFoundByte(ulong match)
...
-                value = default;
-                return false;
+                int sequenceLength = (int)valueSequence.Length;
+                byte* ptr = stackalloc byte[sequenceLength];
+                stackSpan = new Span<byte>(ptr, sequenceLength);
             }
-            return Utf8Parser.TryParse(utf8Unescaped, out value, out int bytesConsumed, 'D') && utf8Unescaped.Length == bytesConsumed;
+
+            valueSequence.CopyTo(stackSpan);
+            span = stackSpan;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs]: We can't out a `span` that has been stackalloc'd.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs [hunk 4]
```diff
@@ -469,8 +469,33 @@ public bool TryGetDateTime(out DateTime value)
...
+            }
+
+            int bytesConsumed;
+
+            if (_stringHasEscaping)
+            {
+                int idx = span.IndexOf(JsonConstants.BackSlash);
+                Debug.Assert(idx != -1);
+
+                Span<byte> utf8Unescaped = stackalloc byte[span.Length];
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: This is an unconstrained stackalloc. It should be constrained, just like the `TryGetGuid`.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs [hunk 5]
```diff
@@ -490,8 +515,33 @@ public bool TryGetDateTimeOffset(out DateTimeOffset value)
...
+                Debug.Assert(written > 0);
+
+                utf8Unescaped = utf8Unescaped.Slice(0, written);
+                Debug.Assert(!utf8Unescaped.IsEmpty);
+
+                return JsonHelpers.TryParseAsISO(utf8Unescaped, out value, out bytesConsumed) && utf8Unescaped.Length == bytesConsumed;
+            }
+
+            Debug.Assert(span.IndexOf(JsonConstants.BackSlash) == -1);
+            return JsonHelpers.TryParseAsISO(span, out value, out bytesConsumed) && span.Length == bytesConsumed;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: Can we do the same length checks here, like `TryGetGuid`?

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs [hunk 6]
```diff
@@ -490,8 +515,33 @@ internal bool TryGetValue(int index, out DateTimeOffset value)
             ReadOnlySpan<byte> data = _utf8Json.Span;
             ReadOnlySpan<byte> segment = data.Slice(row.Location, row.SizeOrLength);
 
-            if (JsonHelpers.TryParseAsISO(segment, out DateTimeOffset tmp, out int bytesConsumed) &&
-                segment.Length == bytesConsumed)
+            DateTimeOffset tmp;
+            int bytesConsumed;
+
+            if (row.HasComplexChildren)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs]: Why do we need this check?

Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs]: This is "needs to be unescaped" :)

Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs]: Oh, lol. Got it. A comment would be helpful :)

**Comment by danmosemsft** [comment]
> This addresses #36202.

@layomia if you use certain words eg `fixes #36202` in your description it will auto close the issue on merge: https://help.github.com/en/articles/closing-issues-using-keywords

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs (hunk 5 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: We'd need to decide on the maximum number of fraction digits allowed. Any thoughts?

File: src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs (hunk 2 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs]: The intention was to copy the contents of the reader's `ValueSequence` to a `span`, then `out` it. Since we can't `out` a `span` that has been stackalloc'd, I've opted to inline this logic in the previously calling methods.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs [hunk 7]
```diff
@@ -490,8 +535,53 @@ public bool TryGetDateTimeOffset(out DateTimeOffset value)
...
+            int bytesConsumed;
+
+            if (_stringHasEscaping)
+            {
+                int idx = span.IndexOf(JsonConstants.BackSlash);
+                Debug.Assert(idx != -1);
+
+                Span<byte> utf8Unescaped = (span.Length <= JsonConstants.StackallocThreshold)
+                    ? stackalloc byte[span.Length]
+                    : new byte[span.Length];
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: Either we should use array pool for the intermediary buffer for unescaping (that is beyond `StackallocThreshold`), or we should constrain the input (similar to `TryGetGuid`) such that stackalloc is fine or for all cases.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs [hunk 8]
```diff
@@ -542,18 +632,25 @@ public bool TryGetGuid(out Guid value)
 
             if (_stringHasEscaping)
             {
-                return JsonReaderHelper.TryGetEscapedGuid(span, out value);
-            }
+                int idx = span.IndexOf(JsonConstants.BackSlash);
+                Debug.Assert(idx != -1);
 
-            Debug.Assert(span.IndexOf(JsonConstants.BackSlash) == -1);
+                Span<byte> utf8Unescaped = stackalloc byte[span.Length];
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: nit: `Debug.Assert` that `span.Length <= JsonConstants.MaximumEscapedGuidLength` to make auditing use of stackalloc easier.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs [hunk 9]
```diff
@@ -490,8 +535,53 @@ public bool TryGetDateTimeOffset(out DateTimeOffset value)
                 throw ThrowHelper.GetInvalidOperationException_ExpectedString(TokenType);
             }
 
-            ReadOnlySpan<byte> span = HasValueSequence ? ValueSequence.ToArray() : ValueSpan;
-            return JsonHelpers.TryParseAsISO(span, out value, out int bytesConsumed) && span.Length == bytesConsumed;
+            ReadOnlySpan<byte> span = ValueSpan;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: Add a comment highlighting that this span potentially points to `stackalloc` space and should not be returned (and only passed down to helpers that don't store it).

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs [hunk 10]
```diff
@@ -490,8 +518,36 @@ internal bool TryGetValue(int index, out DateTimeOffset value)
...
+
+            // Segment needs to be unescaped
+            if (row.HasComplexChildren)
+            {
+                int backslash = segment.IndexOf(JsonConstants.BackSlash);
+                Debug.Assert(backslash != -1);
+
+                Span<byte> segmentUnescaped = (segment.Length <= JsonConstants.StackallocThreshold)
+                    ? stackalloc byte[segment.Length]
+                    : new byte[segment.Length];
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs]: Same here. Consider using array pool rather than allocating a new array or constrain the input size so it "fits" on the stack (like guid).

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs [hunk 11]
```diff
@@ -465,8 +465,36 @@ internal bool TryGetValue(int index, out DateTime value)
             ReadOnlySpan<byte> data = _utf8Json.Span;
             ReadOnlySpan<byte> segment = data.Slice(row.Location, row.SizeOrLength);
 
-            if (JsonHelpers.TryParseAsISO(segment, out DateTime tmp, out int bytesConsumed) &&
-                segment.Length == bytesConsumed)
+            DateTime tmp;
+            int bytesConsumed;
+
+            // Segment needs to be unescaped
+            if (row.HasComplexChildren)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs]: Generally, this path will be uncommon. Is it possible to move the code in this if-block to a helper method instead?

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs (hunk 5 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: I've capped the allowed number of fraction digits to 9, given use cases like https://github.com/dotnet/corefx/issues/36023#issuecomment-472608096.
@JamesNK @bartonjs @rynowak - can you confirm if you are fine with this?

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs (hunk 7 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: We now constrain the input such that stackalloc is fine for all cases.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs [hunk 12]
```diff
@@ -469,8 +469,48 @@ public bool TryGetDateTime(out DateTime value)
...
+
+                if (!JsonReaderHelper.IsValidDateTimeParsingLength(sequenceLength))
+                {
+                    value = default;
+                    return false;
+                }
+
+                Debug.Assert(sequenceLength <= JsonConstants.MaximumEscapedDateTimeOffsetParseLength);
+                Span<byte> stackSpan;
+                // Cannot create a span directly since it gets passed to instance methods on a ref struct.
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: I don't think there are any pass-to-ref-struct cases in this routine (including calls on `this`), so you can simplify things.

```C#
Span<byte> stackSpan = stackalloc byte[sequenceLength];
ValueSequence.CopyTo(stackSpan);
```

You might need to change how `span` is declared to be

```C#
ReadOnlySpan<byte> span = stackalloc byte[0];

if (HasValueSequence)
{
    ...
}
else
{
    span = ValueSpan;
}
```

So that `span` gets the "I'm allowed to hold stackalloc" marker.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs (hunk 5 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: If we do 7 and Java does 9, why not parse out to like 15?  It's still small enough that it's within reason for stackalloc escaping, but large enough that we won't get a "oh, but [some platform] does 10,". (This is just the usual "figure out the biggest "reasonable" value, then add { 10, 25, 50, 100, 100000 }% (depending on what you can accept) of "I don't want to argue" margin)

Random information:
The absolute biggest (under current interpretation of physics) that anyone could legitimately claim is 44 digits ([Planck time](https://en.wikipedia.org/wiki/Planck_time)).  But that seems a bit far. -- The article claims an experiment has measured something in the 19-digit range; following the reference suggests they had 21-digit precision on their measurements.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs (hunk 12 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: Nice. Thanks.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs (hunk 5 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: I think 15 is reasonable for the reason you mentioned. I'll update to this if there are no other thoughts.

File: src/System.Text.Json/src/System/Text/Json/JsonConstants.cs [hunk 13]
```diff
@@ -72,6 +72,11 @@ internal static class JsonConstants
         public const int MaxDateTimeUtcOffsetHours = 14; // The UTC offset portion of a TimeSpan or DateTime can be no more than 14 hours and no less than -14 hours.
         public const int DateTimeNumFractionDigits = 7;  // TimeSpan and DateTime formats allow exactly up to many digits for specifying the fraction after the seconds.
         public const int MaxDateTimeFraction = 9_999_999;  // The largest fraction expressible by TimeSpan and DateTime formats
+        public const int DateTimeParseNumFractionDigits = 15; // The maximum number of fraction digits the Json DateTime parser allows
+        public const int MaximumDateTimeOffsetParseLength = (MaximumFormatDateTimeOffsetLength +
+            (DateTimeParseNumFractionDigits - DateTimeNumFractionDigits)); // Like StandardFormat 'O' for DateTimeOffset, but allowing 8 additional (up to 15) fraction digits.
+        public const int MinimumDateTimeParseLength = 10; // YYYY-MM-DD
+        public const int MaximumEscapedDateTimeOffsetParseLength = MaxExpansionFactorWhileEscaping * MaximumDateTimeOffsetParseLength;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/JsonConstants.cs]: 33 + 8 = 41
41 * 6 = 246

So, we could move `DateTimeParseNumFractionDigits` to 16 and still be < the next power of 2 (i.e .256)

File: src/System.Text.Json/src/System/Text/Json/JsonConstants.cs [hunk 14]
```diff
@@ -72,6 +72,11 @@ internal static class JsonConstants
         public const int MaxDateTimeUtcOffsetHours = 14; // The UTC offset portion of a TimeSpan or DateTime can be no more than 14 hours and no less than -14 hours.
         public const int DateTimeNumFractionDigits = 7;  // TimeSpan and DateTime formats allow exactly up to many digits for specifying the fraction after the seconds.
         public const int MaxDateTimeFraction = 9_999_999;  // The largest fraction expressible by TimeSpan and DateTime formats
+        public const int DateTimeParseNumFractionDigits = 15; // The maximum number of fraction digits the Json DateTime parser allows
+        public const int MaximumDateTimeOffsetParseLength = (MaximumFormatDateTimeOffsetLength +
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/JsonConstants.cs]: Do we need one for just DateTime? i.e. `MaximumDateTimeParseLength` based of  `MaximumFormatDateTimeLength`?

**Comment by layomia** [comment]
I also added (Try)GetGuid to JsonElement.

Coverage for (Try)GetGuid methods is 100%
[report.zip](https://github.com/dotnet/corefx/files/3020208/report.zip)

File: src/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs [hunk 15]
```diff
@@ -226,14 +226,13 @@ private static bool TryParseDateTimeOffset(ReadOnlySpan<byte> source, out DateTi
...
-                while (sourceIndex < source.Length && IsDigit(curByte = source[sourceIndex]))
+                while (sourceIndex < fractionEnd && IsDigit(curByte = source[sourceIndex]))
                 {
-                    int prevFractionTimesTen = fraction * 10;
-
-                    if ((prevFractionTimesTen + (int)(curByte - (uint)'0') <= JsonConstants.MaxDateTimeFraction) && (numDigitsRead < JsonConstants.DateTimeNumFractionDigits))
+                    if (numDigitsRead < JsonConstants.DateTimeNumFractionDigits)
                     {
-                        fraction = prevFractionTimesTen + (int)(curByte - (uint)'0');
+                        fraction = (fraction * 10) + (int)(curByte - (uint)'0');
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs]: Is this change necessary? Would constructing the DateTime with higher precision (beyond 7 digits) change the returned DateTime in anyway?

Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs]: Also, do we need to modify the debug assert here: https://github.com/dotnet/corefx/pull/36347/files#diff-e2e1b5ba65b1a8b77a086eb93e77c2d1R518?

File: src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs [hunk 16]
```diff
@@ -248,27 +249,69 @@ private static int LocateFirstFoundByte(ulong match)
                                                0x02ul << 40 |
                                                0x01ul << 48) + 1;
 
-        public static bool TryGetEscapedGuid(ReadOnlySpan<byte> span, out Guid value)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool IsValidDateTimeParsingLength(int length)
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs]: If DateTime and DateTimeOffset have different "valid lengths", we should split it.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs [hunk 17]
```diff
@@ -469,8 +469,46 @@ public bool TryGetDateTime(out DateTime value)
...
+                int sequenceLength = (int)ValueSequence.Length;
+
+                if (!JsonReaderHelper.IsValidDateTimeParsingLength(sequenceLength))
+                {
+                    value = default;
+                    return false;
+                }
+
+                Debug.Assert(sequenceLength <= JsonConstants.MaximumEscapedDateTimeOffsetParseLength);
+                Span<byte> stackSpan = stackalloc byte[sequenceLength];
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: Is it possible to re-use the outer span and remove `stackSpan all together?
```C#
span = stackalloc byte[sequenceLength];
ValueSequence.CopyTo(stackSpan);
```

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs [hunk 18]
```diff
@@ -469,8 +469,46 @@ public bool TryGetDateTime(out DateTime value)
...
+
+                ValueSequence.CopyTo(stackSpan);
+                span = stackSpan;
+            }
+            else
+            {
+                span = ValueSpan;
+            }
+
+            if (!JsonReaderHelper.IsValidDateTimeParsingLength(span.Length))
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: Move this check into the above else branch. We already validated this for the `HasValueSequence` case.

File: src/System.Text.Json/src/System/Text/Json/JsonConstants.cs (hunk 14 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/JsonConstants.cs]: No because `MaximumFormatDateTimeLength` does not account for a (non-)utc offset which a `DateTime` string could have.
`DateTime` and `DateTimeOffset` have the same valid lengths in our ISO 8601 profile, so we can use the same constants to validate.

File: src/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs (hunk 15 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs]: If we don't cap to 7 digits here, then modify the debug assert, we will diverge from the `Utf8Parser` helpers that we are currently trying to state consistent with (and eventually de-duplicate and rely on: https://github.com/dotnet/corefx/issues/35953).
Also, if we don't enforce a cap, we could overflow at 10 digits.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs (hunk 17 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: No. We can't copy to `ReadOnlySpan` span. Also can't change span to `Span` because we wouldn't be able to assign `ReadOnlySpan` ValueSpan to it.

File: src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs (hunk 16 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.cs]: They have the same "valid lengths".
Now using "InRange".

File: src/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs [hunk 19]
```diff
@@ -836,6 +836,54 @@ public DateTimeOffset GetDateTimeOffset()
...
+        ///   Gets the value of the element as a <see cref="Guid"/>.
+        /// </summary>
+        /// <remarks>
+        ///   This method does not create a Guid representation of values other than JSON strings.
+        /// </remarks>
+        /// <returns>The value of the element as a <see cref="Guid"/>.</returns>
+        /// <exception cref="InvalidOperationException">
+        ///   This value's <see cref="Type"/> is not <see cref="JsonValueType.String"/>.
+        /// </exception>
+        /// <exception cref="ObjectDisposedException">
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs]: Also FormatException

File: src/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs [hunk 20]
```diff
@@ -774,6 +774,9 @@ public bool TryGetDateTime(out DateTime value)
         /// <exception cref="InvalidOperationException">
         ///   This value's <see cref="Type"/> is not <see cref="JsonValueType.String"/>.
         /// </exception>
+        /// <exception cref="FormatException">
+        ///   The value cannot be represented as an <see cref="DateTime"/>.
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs]: nit: as a DateTime

File: src/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs [hunk 21]
```diff
@@ -822,6 +825,9 @@ public bool TryGetDateTimeOffset(out DateTimeOffset value)
         /// <exception cref="InvalidOperationException">
         ///   This value's <see cref="Type"/> is not <see cref="JsonValueType.String"/>.
         /// </exception>
+        /// <exception cref="FormatException">
+        ///   The value cannot be represented as an <see cref="DateTimeOffset"/>.
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonElement.cs]: Same here: an -> a (and for Guid)

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs [hunk 22]
```diff
@@ -469,8 +469,46 @@ public bool TryGetDateTime(out DateTime value)
                 throw ThrowHelper.GetInvalidOperationException_ExpectedString(TokenType);
             }
 
-            ReadOnlySpan<byte> span = HasValueSequence ? ValueSequence.ToArray() : ValueSpan;
-            return JsonHelpers.TryParseAsISO(span, out value, out int bytesConsumed) && span.Length == bytesConsumed;
+            ReadOnlySpan<byte> span = stackalloc byte[0];
+
+            if (HasValueSequence)
+            {
+                int sequenceLength = (int)ValueSequence.Length;
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: nit: Should this be a long?

If `ValueSequence.Length > int.MaxValue32`, we want to return false (the `IsValidDateTimeOffsetParseLength` would catch it).

We should cast to int after `IsValidDateTimeOffsetParseLength` check is done (i.e. when we are stackallocating).

```C#
	{
		long x = (long)int.MaxValue + (long)int.MaxValue + (long)int.MaxValue;
		int sequenceLength = (int)x;
		Console.WriteLine(sequenceLength); // 2147483645
	}

	{
		long x = (long)int.MaxValue + 1L;
		int sequenceLength = (int)x;
		Console.WriteLine(sequenceLength); // -2147483648
	}
```

Otherwise, lets add a comment explaining why this cast is OK since `IsValidDateTimeOffsetParseLength` will return false if `sequenceLength > int.MaxValue` anyway.

Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: Changed to a `long`.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs [hunk 23]
```diff
@@ -469,8 +469,46 @@ public bool TryGetDateTime(out DateTime value)
...
+
+                Debug.Assert(sequenceLength <= JsonConstants.MaximumEscapedDateTimeOffsetParseLength);
+                Span<byte> stackSpan = stackalloc byte[(int)sequenceLength];
+
+                ValueSequence.CopyTo(stackSpan);
+                span = stackSpan;
+            }
+            else
+            {
+                if (!JsonReaderHelper.IsValidDateTimeOffsetParseLength(ValueSpan.Length))
```
Comment (ahsonkhan) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.TryGet.cs]: nit: Maybe we should add an int32 overload of `IsValidDateTimeOffsetParseLength` to avoid forcing the more common int32 code path to be extended to int64.

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs [hunk 24]
```diff
@@ -465,15 +465,24 @@ internal bool TryGetValue(int index, out DateTime value)
...
+            {
+                return JsonReaderHelper.TryGetEscapedDateTime(segment, out value);
+            }
+
+            Debug.Assert(segment.IndexOf(JsonConstants.BackSlash) == -1);
+
             value = default;
-            return false;
+            return (segment.Length <= JsonConstants.MaximumDateTimeOffsetParseLength)
+                && JsonHelpers.TryParseAsISO(segment, out value, out int bytesConsumed)
```
Comment (GrabYourPitchforks) [inline src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs]: Open question: Do we need to prevent the `value` out param from containing an instantiated `DateTime` instance if the `segment.Length == bytesConsumed` check fails and we're about to return false?

(This question applies to a few other places throughout this file.)

File: src/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs (hunk 15 above)
Comment (GrabYourPitchforks) [inline src/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs]: In the PR, you mentioned "Also, if we don't enforce a cap, we could overflow at 10 digits." This is a good observation! It should also exist as a code comment here so that somebody's not tempted to change the value later. :)

File: src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs (hunk 24 above)
Comment (layomia) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs]: Yes, I think we do. I'll follow up with a namespace-wide change following input from @ahsonkhan @bartonjs.

Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs]: I'd say yes: it should emit default(T) when false is returned, and should not emit a semantic value.

Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs]: Agreed.
